src/java/org/apache/cassandra/index/sai/disk/v3/CassandraDiskAnn.java

```diff
@@ -0,0 +1,202 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.index.sai.disk.v3;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.PrimitiveIterator;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import io.github.jbellis.jvector.disk.CachingGraphIndex;
+import io.github.jbellis.jvector.disk.CompressedVectors;
+import io.github.jbellis.jvector.disk.OnDiskGraphIndex;
+import io.github.jbellis.jvector.graph.GraphIndex;
+import io.github.jbellis.jvector.graph.GraphSearcher;
+import io.github.jbellis.jvector.graph.ListRandomAccessVectorValues;
+import io.github.jbellis.jvector.graph.NeighborSimilarity;
+import io.github.jbellis.jvector.graph.SearchResult.NodeScore;
+import io.github.jbellis.jvector.util.Bits;
+import io.github.jbellis.jvector.vector.VectorSimilarityFunction;
+import org.apache.cassandra.index.sai.IndexContext;
+import org.apache.cassandra.index.sai.QueryContext;
+import org.apache.cassandra.index.sai.disk.format.IndexComponent;
+import org.apache.cassandra.index.sai.disk.v1.PerIndexFiles;
+import org.apache.cassandra.index.sai.disk.v1.SegmentMetadata;
+import org.apache.cassandra.index.sai.disk.v1.postings.ReorderingPostingList;
+import org.apache.cassandra.index.sai.disk.vector.CassandraOnHeapGraph;
+import org.apache.cassandra.index.sai.disk.vector.JVectorLuceneOnDiskGraph;
+import org.apache.cassandra.index.sai.disk.vector.OnDiskOrdinalsMap;
+
+public class CassandraDiskAnn implements JVectorLuceneOnDiskGraph, AutoCloseable
+{
+    private static final Logger logger = Logger.getLogger(CassandraDiskAnn.class.getName());
+
+    private final OnDiskOrdinalsMap ordinalsMap;
+    private final CachingGraphIndex graph;
+    private final VectorSimilarityFunction similarityFunction;
+
+    // only one of these will be not null
+    private final CompressedVectors compressedVectors;
+    private final ListRandomAccessVectorValues uncompressedVectors;
+
+    public CassandraDiskAnn(SegmentMetadata.ComponentMetadataMap componentMetadatas, PerIndexFiles indexFiles, IndexContext context) throws IOException
+    {
+        similarityFunction = context.getIndexWriterConfig().getSimilarityFunction();
+
+        SegmentMetadata.ComponentMetadata termsMetadata = componentMetadatas.get(IndexComponent.TERMS_DATA);
+        graph = new CachingGraphIndex(new OnDiskGraphIndex<>(() -> indexFiles.termsData().createReader(), termsMetadata.offset));
```

Comment:
We should be making a shared copy of the `FileHandle` here and closing it when we are finished with it. I hit an issue with this in the CEP because the OS code changed such that the shared copy was made when the file handle was read from the `PerIndexFiles`. I got round that by introducing a local `TermsReaderSupplier` class that correctly closed the file handle.

I'm not 100% sure of the implications of leaving this as is because the only issue would be if another user of the file handle closed it and the handle was released. I'm not sure that can happen and this is the only user of it.

Reply:
This line creates a Supplier, it does not open a Reader/FileHandle.  The Supplier is invoked when getView is called, at which point it's the caller of getView's responsibility to call close().

Reply:
Yes but it closes the `RandomAccessReader` but not the `FileHandle`.

Reply:
Wouldn't someone else closing it (before it's compacted away) be a bug?

Reply:
Yes,  it would, but the whole design of the `PerIndexFiles` is predicated on the user of a file calling `sharedCopy` and then releasing it when they have finished with it. This is why we changed the OS code to call `sharedCopy` in `getFile`. 

I'm not saying what you are doing is wrong but it is dangerous, for the future, if a user of that `FileHandle` is not aware that `sharedCopy` has not been called and releases it.

Reply:
Maybe I'm being pedantic too, but it doesn't sit well with me to make defensive copies against something that shouldn't happen.  What would happen if we documented PerIndexFiles as intended to live for the duration of the index, don't close it?

Alternatively, if you're sure the current design is the correct one, I think we should change PIF to hold a FileHandleFactory and not an actual FileHandle to avoid confusion.

Reply:
TBH, I don't think the current solution is the correct one at all. The PerIndexFiles do remain in place for the duration of the index so I'm not entirely sure why we do the sharedCopy.

Reply:
I'm saying it because that is the way that we have done it up to now. 

I did a bit of digging into this to try an establish exactly why we need to do this. The closest I can come to a reasonable explanation is that we need to hold a copy of the `FileHandle` to avoid it getting closed when the underlying sstable index is released by a sstable lifecycle change. The `PerIndexFiles` are, effectively, owned by the `SSTableIndex` so will get closed when the `SSTableIndex` is released. This will close all the file handles it owns.

Reply:
Okay, latest commit updates PIF to do the right thing by default.